tests/suspend_nodes.py

```diff
@@ -0,0 +1,195 @@
+
+
+@reqs.description("Suspend nodes")
+def test_suspend_nodes(network, args, nodes_to_keep):
+    cur_primary, _ = network.find_primary()
+
+    # first timer determines after how many seconds each node will be suspended
+    timeouts = []
+    for i, node in enumerate(nodes_to_keep):
+        # if pbft suspend half of them including the primary
+        if i % 2 != 0 and args.consensus == "pbft":
+            continue
+        LOG.success(f"Will suspend node with id {node.node_id}")
+        t = random.uniform(0, 2)
+        LOG.info(f"Initial timer for node {node.node_id} is {t} seconds...")
+        timeouts.append((t, node))
+
+    for t, node in timeouts:
+        suspend_time = (
+            args.pbft_view_change_timeout / 1000
+            if args.consensus == "pbft"
+            else args.raft_election_timeout / 1000
+        )
+        if node.node_id == cur_primary.node_id and args.consensus == "pbft":
+            # if pbft suspend the primary for more than twice the election timeout
+            # in order to make sure view changes will be triggered
+            suspend_time = 2.5 * suspend_time
+        tm = Timer(t, timeout_handler, args=[node, True, suspend_time])
+        tm.start()
+    return network
+
+
+def run(args):
+    hosts = ["localhost", "localhost", "localhost"]
+
+    with infra.ccf.network(
+        hosts, args.binary_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        network.start_and_join(args)
+        original_nodes = network.get_joined_nodes()
+        term_info = {}
+        update_term_info(network, term_info)
+        txs = app.LoggingTxs()
+
+        app.test_run_txs(
+            network=network, args=args, num_txs=TOTAL_REQUESTS, logging_app=txs
+        )
+        update_term_info(network, term_info)
+
+        nodes_to_kill = [network.find_any_backup()]
+        nodes_to_keep = [n for n in original_nodes if n not in nodes_to_kill]
+
+        # check that a new node can catch up after all the requests
+        late_joiner = network.create_and_trust_node(args.package, "localhost", args)
+        nodes_to_keep.append(late_joiner)
+
+        # some requests to be processed while the late joiner catches up
+        # (no strict checking that these requests are actually being processed simultaneously with the node catchup)
+        app.test_run_txs(
+            network=network,
+            args=args,
+            num_txs=int(TOTAL_REQUESTS / 2),
+            nodes=original_nodes,  # doesn't contain late joiner
+            verify=False,  # will try to verify for late joiner and it might not be ready yet
+            logging_app=txs,
+        )
+
+        wait_for_late_joiner(original_nodes[0], late_joiner)
+
+        # kill the old node(s) and ensure we are still making progress
+        for backup_to_retire in nodes_to_kill:
+            LOG.success(f"Stopping node {backup_to_retire.node_id}")
```

Comment:
Minor: I don't think we need this log as it gets printed by the node on shutdown anyway?

Reply:
The hostname is printed by the node shutdown, it's just a bit easier to see in the logs this way which node was actually stopped